### PR TITLE
[#173960515] Add ability to create resolution

### PR DESCRIPTION
**Pull request by ShaidullinArtur**
- [Story 173960515](https://www.pivotaltracker.com/story/show/173960515)

### CODE REVIEW

File: app/controllers/communitarian/resolutions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    requires_plugin Communitarian
+
+    before_action :ensure_logged_in
+
+    def create
+      manager = NewPostManager.new(current_user, permitted_params)
+
+      result = manager.perform
+      json = serialize_data(result, NewPostResultSerializer, root: false)
+      render json: json, status: result.success? ? 200 : 422
```
Comment (DmitryBarskov) [inline app/controllers/communitarian/resolutions_controller.rb]:
```suggestion
      render json: json, status: result.success? ? :created : :unprocessable_entity
```

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    let closeDate = moment.tz("america_new_york").set({ hours: 17, minutes: 0, seconds: 0, millisecond: 0 });
+    const currentWeekday = closeDate.weekday();
+    const closeWeekDay = 6;
+
+    // choose next saturday if current work week has been finished
+    if (currentWeekday > 0 && currentWeekday < 6) {
+      closeDate.isoWeekday(closeWeekDay);
+    } else {
+      closeDate.add(1, "week").isoWeekday(closeWeekDay);
+    }
```
Comment (DmitryBarskov) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]:
```suggestion
    const isWeekEnd = (weekday) => weekday > 0 && weekday < 6;

    if (isWeekEnd(currentWeekday)) {
      closeDate.add(1, "week");
    }
    closeDate.isoWeekday(closeWeekDay);
```

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    });
+
+    return length;
+  },
+
+  @discourseComputed("pollOptions")
+  pollOutput(pollOptions) {
+    let pollHeader = "[poll type=regular results=always chartType=bar";
+    let output = "";
+    let closeDate = moment.tz("america_new_york").set({ hours: 17, minutes: 0, seconds: 0, millisecond: 0 });
```
Comment (DmitryBarskov) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]: Would you like to have a function for choosing next close date?

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    pollHeader += "]";
+    output += `${pollHeader}\n`;
+
+    if (pollOptions.length > 0) {
+      pollOptions.split("\n").forEach(option => {
+        if (option.length !== 0) output += `* ${option}\n`;
+      });
+      output += `* ${I18n.t("communitarian.resolution.ui_builder.poll_options.close_option")}\n`
+    }
```
Comment (DmitryBarskov) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]: Would you like to have function `parsePollOptions` for that?

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    return EmberObject.create(options);
+  },
+
+  @discourseComputed("pollOptionsCount", "title", "loading")
+  disabledButton(pollOptionsCount, title, loading) {
+    if (loading) {
+      return true;
+    } else {
+      return (pollOptionsCount < 1 || title.length > this.siteSettings.max_topic_title_length);
+    }
```
Comment (DmitryBarskov) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]:
```suggestion
    return loading ||
           pollOptionsCount < 1 ||
           title.length > this.siteSettings.max_topic_title_length;
```

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+export default Controller.extend({
+  init() {
+    this._super(...arguments);
+    this._setupPoll();
+  },
+
+  @discourseComputed("pollOptions")
+  pollOptionsCount(pollOptions) {
+    if (pollOptions.length === 0) return 0;
```
Comment (DmitryBarskov) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]: Do we really need this check? If `pollOptions` is 0, the code below also returns 0.

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    let length = 0;
+
+    pollOptions.split("\n").forEach(option => {
+      if (option.length !== 0) length += 1;
+    });
+
+    return length;
```
Comment (DmitryBarskov) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]:
```suggestion
    let notEmpty = Boolean;
    return pollOptions.split("\n").filter(notEmpty).length;
```
Or
```suggestion
    let countNotEmpty = (count, s) => count + (s.length === 0 ? 0 : 1);
    return pollOptions.split("\n").reduce(countNotEmpty, 0);
```

**Review by DmitryBarskov — APPROVED** [review]
LGFM, except for the comments.

File: config/locales/client.en.yml [hunk 8]
```diff
@@ -1,4 +1,16 @@
...
+          form_title: New Resolution
+          create: Create Resolution
+          title:
+            label: Type title, or paste a link here
+          help:
+            options_count: Enter at least 1 option
+          poll_options:
+            label: Enter one poll option per line
+            close_option: Close the poll
+        error_while_creating: "Sorry, there was an error creating the resolution."
```
Comment (Vadim-Kurnatovskiy) [inline config/locales/client.en.yml]: remove unnecessary quotes from error_while_creating message

File: config/routes.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+require_dependency "communitarian_constraint"
+
+Communitarian::Engine.routes.draw do
+  post '/resolutions' => 'resolutions#create', constraints: CommunitarianConstraint.new
```
Comment (Vadim-Kurnatovskiy) [inline config/routes.rb]: Use double quotes instead of single quotes.

File: assets/javascripts/controllers/resolution-ui-builder.js.es6 [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+      titleMaxLength: this.siteSettings.max_topic_title_length,
+      loading: false,
+      typingTime: 0,
+      firstOpenedTimestamp: new Date(),
+      category: window.location.pathname.match(/c\/.*\/(.*)$/)[1],
+    });
+  },
+
+  _closeDate() {
+    let closeDate = moment.tz("america_new_york").set({ hours: 17, minutes: 0, seconds: 0, millisecond: 0 });
```
Comment (Vadim-Kurnatovskiy) [inline assets/javascripts/controllers/resolution-ui-builder.js.es6]: Use `America/New_York` format instead of `america_new_york`

**Review by Vadim-Kurnatovskiy — APPROVED** [review]
LGFM except comments

**Pull request merged**